# IC-UFPA-HOMEWORK sobre SVMs

Filipe Corrêa da Silva - 202006840020

    1 - A figura abaixo mostra os exemplos de treino e as regiões de decisão de uma SVM (ou SVC, de “support vector classifier”) para um problema binário (duas classes onde o vetor x de “features” de entrada tem dois parâmetros (weight e length). Os pontos em azul e vermelho, representam os exemplos de cada classe. As regiões em azul e vermelho representam as regiões nas quais o classificador irá prever uma classe azul e vermelha, respectivamente. Olhando para as figuras, informe:
    
    a) Quantos erros cada SVM possui no conjunto de treino?

LinearSVC (linear kernel) -> 2 erros

SVC with linear kernel -> 1 erro

SVC with RBF kernel -> 1 erro

SVC with polynomial (degree 3) kernel -> 0 erros

    b) Qual a que lhe parece ser a melhor SVM? Por que?

A melhor SVM me parece ser a última (SVC with polynomial (degree 3) kernel),  pois ela não apresenta erros no seu conjunto de treino.    

    2 - A regularização é uma importante ferramenta para controlar a “complexidade do modelo” durante a fase de seu projeto. No projeto de uma SVM, o hiperparâmetro C é usado para efetuar a regularização do modelo na etapa de seleção do modelo, ou seja, quando os hiperparâmetros são variados e observamos o impacto através do erro de classificação no conjunto de validação. Suponha que você se encontra em uma situação onde está descontente com o atual resultado de sua SVM e observa que o número de vetores de suporte está relativamente pequeno. Você quer potencialmente aumentar o número de vetores de suporte, fazendo com que sua nova SVM seja um modelo mais “complexo”. Para isso, você deve aumentar ou diminuir o parâmetro “C” da classe SVC do scikit-learn?    

Com o parâmetro C aumentado, a SVM prioriza a classificação correta dos dados de treinamento, mesmo que a margem seja menor. Isso resulta em mais vetores de suporte, pois o modelo se ajusta mais aos dados, incluindo pontos próximos ao hiperplano. Logo, se o número de vetores de suporte está muito pequeno e se deseja aumentar a complexidade do modelo, aumentar C fará com que a SVM seja mais flexível, incluindo mais vetores de suporte e reduzindo a margem.

    3 - Um classificador usa vetores de entrada com dimensão K=5 “features”. Após treinar uma SVM linear, o número de vetores de suporte foi de 450 exemplos. Neste caso, calcular o kernel linear corresponde a um produto interno entre dois vetores de dimensão K=5, cada, o que requer K multiplicações e K-1 adições. Estime o fator F = Coriginal / Cperceptron de redução do custo computacional da etapa de teste ao converter esta SVM linear para um perceptron. Assuma que os custos Coriginal Cperceptron correspondem ao número de multiplicações e adições usando-se respectivamente, a SVM original com os 450 vetores de suporte e após sua conversão para perceptron.

Cálculo do Custo da SVM Original:
Em uma SVM linear com 450 vetores de suporte (SV), a classificação de uma nova entrada x é dada por:

$$
f(\mathbf{x}) = \text{sgn}\left( \sum_{i=1}^{450} \alpha_i y_i \langle \mathbf{x}_i, \mathbf{x} \rangle + b \right)
$$

Cálculo do Produto Interno (<xi, x>):
Para cada vetor de suporte xi, o produto interno com x requer:
- **5 multiplicações** (uma para cada dimensão),
- **4 adições** (para somar os resultados das multiplicações).

Como há **450 vetores de suporte**, o custo total para calcular todos os produtos internos é:

$$
450 \times (5 \text{ multiplicações} + 4 \text{ adições}) = 450 \times 9 = 4050 \text{ operações}.
$$

Cálculo da Soma Ponderada e Bias:
1. Multiplicar cada $ <xi, x> $ por $ \alpha_i y_i $ (**450 multiplicações**),
2. Somar todos os termos (**449 adições**),
3. Adicionar o bias \( b \) (**1 adição**).

**Total adicional:**

$$
450 + 449 + 1 = 900 \text{ operações}.
$$

Custo Total $ C_{\text{original}} $
O custo computacional total da SVM é:

$$
C_{\text{original}} = 4050 + 900 = 4950 \text{ operações}.
$$

**2. Custo do Perceptron $ C_{\text{perceptron}} $**
Ao converter a SVM linear em um perceptron, o hiperplano de decisão é dado por:

$$
f(\mathbf{x}) = \text{sgn}\left( \langle \mathbf{w}, \mathbf{x} \rangle + b \right)
$$

onde

$$
\mathbf{w} = \sum_{i=1}^{450} \alpha_i y_i \mathbf{x}_i
$$

é o vetor de pesos.

Cálculo do Produto Interno $ (<w, x>) $
- O vetor w tem dimensão $ K=5 $, assim como x.
- O produto interno requer:
  - **5 multiplicações** (uma para cada dimensão),
  - **4 adições** (para somar os resultados).

**Adição do Bias**
- **1 adição** para incluir \( b \).

**Custo Total (\( C_{\text{perceptron}} \))**
$$
C_{\text{perceptron}} = 5 + 4 + 1 = 10 \text{ operações}.
$$

**3. Fator de Redução (\( F \))**
$$
F = \frac{C_{\text{original}}}{C_{\text{perceptron}}} = \frac{4950}{10} = 495.
$$

O fator de redução do custo computacional é **495**.  
Ou seja, o perceptron é **495 vezes mais eficiente** que a SVM original no teste.

    4 - Interpretação do resultado de projeto de SVMs.
    O treinamento de uma SVM linear usando a classe SVC do sklearn retornou o resultado:
    
    svm.n_support_= [1 2] # número de vetores de suporte por classe
    svm.support_vectors_= [[ 1. 4.] # vetores de suporte
                          [-2. 3.]
                          [-2. -5.]]
    svm.dual_coef_= [[-0.5 -0.3 0.8]] # valores de \lambda
    svc.intercept_= [-2] # “bias”
    Observe que esta questão não usa o conjunto de treino indicado, mas outro. Pede-se:

    a) A função de decisão para essa SVM de acordo com a fórmula geral:

A função de decisão de uma SVM é dada por:  

$$
f(\mathbf{z}) = \left( \sum_{n=0}^{N-1} \lambda_n K(\mathbf{z}, \mathbf{x}_n) \right) + b
$$

onde:  
- $ \lambda_n \ $ são os coeficientes duais (fornecidos por dual_coef_),  
- $ K(\mathbf{z}, \mathbf{x}_n) $ é o kernel linear (produto interno $ \langle \mathbf{z}, \mathbf{x}_n \rangle $).  

Substituindo os valores:  

$$
f(\mathbf{z}) = -0.5 \cdot \langle \mathbf{z}, [1, 4] \rangle - 0.3 \cdot \langle \mathbf{z}, [-2, 3] \rangle + 0.8 \cdot \langle \mathbf{z}, [-2, -5] \rangle - 2.
$$  


    b) A função de decisão desta SVM quando escrita como um perceptron f(z)=<z,w>+b.

**b) Função de decisão como um perceptron**  
A SVM linear pode ser reescrita como um perceptron:  

$$
f(\mathbf{z}) = \langle \mathbf{z}, \mathbf{w} \rangle + b
$$

onde $ \mathbf{w} = \sum_{n=0}^{N-1} \lambda_n y_n \mathbf{x}_n $.  

Calculando $ \mathbf{w} $:  

$$
\mathbf{w} = -0.5 \cdot [1, 4] - 0.3 \cdot [-2, 3] + 0.8 \cdot [-2, -5].
$$

$$
\mathbf{w} = [-0.5 \cdot 1 - 0.3 \cdot (-2) + 0.8 \cdot (-2), \ -0.5 \cdot 4 - 0.3 \cdot 3 + 0.8 \cdot (-5)].
$$

$$
\mathbf{w} = [-0.5 + 0.6 - 1.6, \ -2 - 0.9 - 4] = [-1.5, \ -6.9].
$$  

Portanto, a função do perceptron é:  
$$
f(\mathbf{z}) = <z[-1.5 - 6.9]> - 2.
$$

$$
f(\mathbf{z}) = -1.5 z_1 - 6.9 z_2 - 2.
$$  


    c) A saída da função de decisão quando o vetor de entrada é z=[0, 0] e o respectivo rótulo predito y assumindo-se que y = I( f(z) > 0 ), onde I(.) é a função “indicador”, que é 1 se o argumento é verdadeiro, ou 0 em caso contrário.

Substituindo $ \mathbf{z} = [0, 0] $ na função do perceptron:  

$$
f([0, 0]) = -1.5 \cdot 0 - 6.9 \cdot 0 - 2 = -2.
$$  

O rótulo predito é:  

$$
y = \mathbb{I}(f(\mathbf{z}) > 0) = \mathbb{I}(-2 > 0) = 0.
$$  



    5 - Interpretação do resultado de projeto de SVMs.
    O código ak_svm_prova1.py e o conjunto de treino em dataset_train.txt foram usados para
    gerar a figura e o log (saída de texto) abaixo. Em suma, foram treinadas 4 SVMs, duas com
    kernel linear (SVMs 1 e 2) e outras duas com kernel não-linear (SVMs 3 e 4).
    
    Pergunta-se:
    
    a) Quais as expressões para as SVMs 3 e 4 (não-lineares) e para a SVM 2 (linear) de acordo
    com a nomenclatura abaixo (escreva a “fórmula” da função de decisão de cada SVM
    indicando o valor do bias “b”, etc., considerando que z e xn são vetores já normalizados (com
    o primeiro elemento dividido por 1000).
    
**Formato geral da função de decisão**:

$$
f(\mathbf{z}) = \left( \sum_{n=0}^{N-1} \lambda_n K(\mathbf{z}, \mathbf{x}_n) \right) + b
$$

**SVM 2 (Linear)**
- **Kernel**: Linear $ K(\mathbf{z}, \mathbf{x}_n) = \langle \mathbf{z}, \mathbf{x}_n \rangle $.
- **Vetores de suporte**:

$$
\mathbf{x}_0 = [0, -4], \quad \mathbf{x}_1 = [-1, 2], \quad \mathbf{x}_5 = [-2, -2]
$$

- **Coeficientes duais**:

$$
\lambda = [-0.46, -0.28, 0.74]
$$

- **Bias**: \( b = -1.8 \).

**Função de decisão**:

$$
f(\mathbf{z}) = -0.46 \cdot \langle \mathbf{z}, [0, -4] \rangle - 0.28 \cdot \langle \mathbf{z}, [-1, 2] \rangle + 0.74 \cdot \langle \mathbf{z}, [-2, -2] \rangle - 1.8
$$

**SVM 3 (RBF Kernel)**
- **Kernel**: RBF $ K(\mathbf{z}, \mathbf{x}_n) = \exp(-\gamma \|\mathbf{z} - \mathbf{x}_n\|^2) $ , com  $ \gamma = 0.7 $.
- **Vetores de suporte**: Todos os 6 exemplos de treino (índices 0 a 5).
- **Coeficientes duais**:

$$
\lambda = [-0.91, -0.91, -0.91, 0.87, 0.87, 1.0]
$$

- **Bias**: \( b = -0.08 \).

**Função de decisão**:

$$
f(\mathbf{z}) = \sum_{n=0}^{5} \lambda_n \cdot \exp\left(-0.7 \|\mathbf{z} - \mathbf{x}_n\|^2\right) - 0.08
$$

**SVM 4 (Polinomial Kernel)**
- **Kernel**: Polinomial $ K(\mathbf{z}, \mathbf{x}_n) = (\gamma \langle \mathbf{z}, \mathbf{x}_n \rangle + \text{coef}_0)^\text{degree} $ , com $  \gamma = \text{auto} ,  \text{degree} = 3 ,  \text{coef}_0 = 0  $.
- **Vetores de suporte**: Índices 0, 1, 5.
- **Coeficientes duais**:

$$
\lambda = [-0.008, -0.03, 0.04]
$$

- **Bias**: \( b = -1.03 \).

**Função de decisão**:

$$
f(\mathbf{z}) = \sum_{n \in \{0,1,5\}} \lambda_n \cdot (\gamma \langle \mathbf{z}, \mathbf{x}_n \rangle)^3 - 1.03731897
$$

    b) Quais as expressões para as SVMs 1 e 2 (lineares) quando escritas como perceptrons?

**SVM 1 (LinearSVC)**
- **Vetor de pesos**:

$$
\mathbf{w} = [-0.68, -0.10]
$$

- **Bias**: \( b = -0.99556119 \).

**Função do perceptron**:

$$
f(\mathbf{z}) = -0.68 z_1 - 0.10 z_2 - 0.99556119
$$

**SVM 2 (SVC com Kernel Linear)**
- **Vetor de pesos**:

$$
\mathbf{w} = [-1.19, -0.19]
$$

- **Bias**: \( b = -1.8 \).

**Função do perceptron**:

$$
f(\mathbf{z}) = -1.19 z_1 - 0.19 z_2 - 1.8
$$

    
    c) Suponha que para a SVM 2 (linear), apenas as informações abaixo fossem fornecidas.
    Com elas, você já pôde escrever essa SVM no formato geral de uma SVM, como no item a).
    
1. **Calcular $ \mathbf{w} $**:

$$
\mathbf{w} = \sum_{n} \lambda_n \mathbf{x}_n = -0.45994152 \cdot [0, -4] - 0.27992202 \cdot [-1, 2] + 0.73986354 \cdot [-2, -2]
$$

Resultado:

$$
\mathbf{w} = [-1.19, -0.19]
$$

2. **Manter o bias**: \( b = -1.8 \).

**Economia computacional**:
- **SVM original**: **11 operações**.
- **Perceptron**: **3 operações**.
- **Redução**: $ \frac{11}{3} \approx 3.67\times $ mais eficiente.

  
    Explique agora com clareza quais os passos que você adotaria para converter essa SVM linear
    em um perceptron como descrito no item b) e indique em termos do número de multiplicações
    e adições, qual economia no custo computacional que a implementação como perceptron
    alcança.
    
    svm.n_support_= [2 1]
    svm.support_= [0 1 5]
    svm.support_vectors_= [[ 0. -4.]
    [-1. 2.]
    [-2. -2.]]
    svm.dual_coef_= [[-0.45994152 -0.27992202 0.73986354]]
    svc.intercept_= [-1.79954513]
    
    d) Para as SVMs 3 e 4 (não-lineares), indique:
    d.1) qual o número total de vetores de suporte (SVs), considerando todas as classes.

- **Número total de vetores de suporte (SVs)**:
  - **SVM 3 (RBF)**: 6 SVs (índices 0 a 5).
  - **SVM 4 (Polinomial)**: 3 SVs (índices 0, 1, 5).
    

    d.2) Quais os índices desses SVs no conjunto de treino e os respectivos valores dos “lambdas” (coeficientes duais)
    

**SVM 3:**  
Coeficientes duais $ \lambda $:

$$
\lambda = [-0.91722233, -0.91351914, -0.91300432, 0.87185969, 0.8718861, 1.0]
$$

**SVM 4:**  
Coeficientes duais $ \lambda\ $:

$$
\lambda = [-0.00887134, -0.03133903, 0.04021037]
$$

    d.3) Qual o valor do termo independente b chamado de “bias” ou “intercept_”.
    

- **Valores do bias \( b \)**:
  - **SVM 3**: \( b = -0.08676121 \).
  - **SVM 4**: \( b = -1.03731897 \).


    e) Considerando as saídas da SVM abaixo para o conjunto de treino, em qual dos exemplos
    de treino esta SVM está menos “confiante” (assumindo que esses números são “confidence
    scores”) em sua decisão e qual é a classe que esta SVM prediz para este exemplo?
    svm.decision_function(X)= [-1.00027976 -1.00027976 -0.99977173 1.00010297
    1.00022828 0.90993821]


**Saída da SVM 3 (RBF)**:

$$
\text{decision\_function}(X) = [-1.00027976, -1.00027976, -0.99977173, 1.00010297, 1.00022828, 0.90993821]
$$

- **Menor confiança**: Índice 2 (\(-0.99977173\)).
- **Classe predita**: $ \mathbb{I}(f(\mathbf{z}) > 0) = 0 $.

    6) Use o scikit-learn para treinar uma SVM com kernel linear, usando hiperparâmetro: C=1.
    Neste caso use a versão normalizada dos dados, com os fatores de normalização projetados
    com base no conjunto de treino, para que este conjunto tenha média 0 e variância 1.
    1) Treine a SVM com o conjunto de treino indicado no arquivo dataset_train.txt.
    2) Indique o desempenho do modelo SVM no conjunto de teste dataset_test.txt.
    3) Converta esta SVM para um perceptron, estime o custo computacional durante a fase de
    teste da SVM na forma original com sua versão implementada como perceptron. Indique se
    há algum ganho computacional em termos de memória e número de operações ao se
    converter essa SVM para um perceptron.

In [ ]:
import numpy as np
from sklearn.svm import SVC
from sklearn.preprocessing import StandardScaler

# 1. Carregar os dados de treino e teste
def load_data(train_path, test_path):
    X_train = np.loadtxt(train_path, delimiter=',', usecols=(0, 1))
    y_train = np.loadtxt(train_path, delimiter=',', usecols=2)
    X_test = np.loadtxt(test_path, delimiter=',', usecols=(0, 1))
    y_test = np.loadtxt(test_path, delimiter=',', usecols=2)
    return X_train, y_train, X_test, y_test

# 2. Normalizar os dados (média=0, variância=1)
def normalize_data(X_train, X_test):
    scaler = StandardScaler()
    X_train_norm = scaler.fit_transform(X_train)
    X_test_norm = scaler.transform(X_test)
    return X_train_norm, X_test_norm, scaler

# 3. Treinar SVM linear
def train_svm(X_train, y_train, C=1):
    svm = SVC(kernel='linear', C=C)
    svm.fit(X_train, y_train)
    return svm

# 4. Converter SVM para perceptron
def svm_to_perceptron(svm):
    w = svm.coef_[0]  # Vetor de pesos (já considera lambda * y)
    b = svm.intercept_[0]  # Bias
    return w, b

# 5. Calcular custo computacional
def compute_cost(svm, w):
    n_sv = sum(svm.n_support_)  # Total de vetores de suporte
    k = X_train.shape[1]  # Dimensão dos dados (features)

    # Custo da SVM original
    svm_ops = n_sv * (k + 1)  # Produtos internos + soma ponderada
    svm_memory = n_sv * k + n_sv + 1  # SVs + lambdas + bias

    # Custo do perceptron
    perceptron_ops = k + 1  # Produto interno + bias
    perceptron_memory = k + 1  # w + b

    return {
        'svm_ops': svm_ops,
        'svm_memory': svm_memory,
        'perceptron_ops': perceptron_ops,
        'perceptron_memory': perceptron_memory,
        'speedup': svm_ops / perceptron_ops,
        'memory_reduction': 1 - (perceptron_memory / svm_memory)
    }

# Pipeline principal
if __name__ == "__main__":
    # Carregar e normalizar dados
    X_train, y_train, X_test, y_test = load_data('dataset_train.txt', 'dataset_test.txt')
    X_train_norm, X_test_norm, scaler = normalize_data(X_train, X_test)

    # Treinar SVM
    svm = train_svm(X_train_norm, y_train, C=1)
    print(f"Acurácia no teste: {svm.score(X_test_norm, y_test):.4f}")
    print(f"Número de vetores de suporte: {svm.n_support_}")

    # Converter para perceptron
    w, b = svm_to_perceptron(svm)
    print(f"\nPerceptron: f(z) = {w[0]:.6f} * z1 + {w[1]:.6f} * z2 + {b:.6f}")

    # Calcular custos
    costs = compute_cost(svm, w)
    print("\nCusto computacional:")
    print(f"- SVM original: {costs['svm_ops']} operações por classificação")
    print(f"- Perceptron: {costs['perceptron_ops']} operações por classificação")
    print(f"- Ganho de velocidade: {costs['speedup']:.1f}x")
    print(f"- Memória SVM: {costs['svm_memory']} valores armazenados")
    print(f"- Memória Perceptron: {costs['perceptron_memory']} valores armazenados")
    print(f"- Redução de memória: {costs['memory_reduction']*100:.1f}%")

Acurácia no teste: 1.0000
Número de vetores de suporte: [1 2]

Perceptron: f(z) = -1.117775 * z1 + -0.355624 * z2 + 0.144827

Custo computacional:
- SVM original: 9 operações por classificação
- Perceptron: 3 operações por classificação
- Ganho de velocidade: 3.0x
- Memória SVM: 10 valores armazenados
- Memória Perceptron: 3 valores armazenados
- Redução de memória: 70.0%


Os resultados apresentados mostram a conversão de uma SVM linear para um perceptron e os ganhos computacionais obtidos com essa transformação. A SVM original alcançou acurácia perfeita (1.0000) no conjunto de teste, utilizando três vetores de suporte (1 da primeira classe e 2 da segunda classe). Após a conversão para perceptron, a função de decisão passou a ser expressa pela fórmula linear $ f(z) = -1.117775 \cdot z_1 - 0.355624 \cdot z_2 + 0.144827 $, onde \( z_1 \) e \( z_2 \) são as features do vetor de entrada e o termo constante representa o *bias*.

Em termos de eficiência computacional, a conversão para perceptron trouxe benefícios significativos. A SVM original exigia 9 operações por classificação, envolvendo o cálculo de produtos internos com todos os vetores de suporte e uma soma ponderada. Já o perceptron reduz esse custo para apenas 3 operações por classificação, correspondendo a um produto interno simples entre o vetor de entrada e os pesos do perceptron, seguido da adição do *bias*. Isso representa um ganho de velocidade de 3 vezes, tornando o processo de classificação mais rápido e adequado para aplicações em tempo real ou sistemas com recursos limitados.

Além da vantagem em velocidade, a redução no uso de memória também é notável. A SVM original precisava armazenar 10 valores, incluindo os vetores de suporte e seus coeficientes, enquanto o perceptron requer apenas 3 valores (os dois pesos e o *bias*), resultando em uma redução de 70% no consumo de memória. Essa otimização é especialmente útil em dispositivos com restrições de hardware, como microcontroladores ou sistemas embarcados, onde a eficiência de recursos é crítica. Em resumo, a conversão para perceptron mantém a acurácia do modelo original, mas com ganhos expressivos em desempenho e economia de memória.

    7) Sobre o projeto de SVMs:
    a) Use o scikit-learn (exemplificado no código ak_svm.py) para treinar uma SVM com kernel
    Gaussiano (também chamado de “RBF”), usando o conjunto de validação para fazer o
    “tuning” (otimização) de dois hiperparâmetros: “C” e “gamma”, avaliando os valores:
    C = 0.01
    C = 1
    C = 100
    gamma = 0.5
    gamma = 1
    Em suma:
    1) Treine as SVMs com o conjunto de treino variando C e gamma;
    2) Observe o resultado usando o conjunto de validação;
    3) Indique qual SVM apresenta melhor performance no conjunto de validação.
    Obs: O produto Cartesiano das opções de hiperparâmetros irá exigir 3 x 2 = 6 treinos de
    SVMs.
    b) Descreva totalmente (por extenso) os parâmetros que compõem a fórmula final do
    classificador SVM escolhido, indicando:
    b1) Os parâmetros escolhidos do kernel (gamma neste caso, visto que C é apenas um
    hiperparâmetro, usado no treino mas não no teste)
    b2) O número de vetores de suporte (SVs)
    b3) Seus índices no conjunto de treino
    b4) Os respectivos valores dos “lambdas” (ou “dual_coef_” de coeficientes duais) e
    b5) O termo independente b chamado de “bias” ou “intercept_”. Em outras palavras, indique
    os números que serão usados na função de decisão f, que pode ser escrita abaixo, onde n é o
    índice do SV no conjunto de treino:
    ou como na documentação do scikit-learn em https://scikit-
    learn.org/stable/modules/svm.html#svm-mathematical-formulation, onde li = yi ai, onde yi é
    o “label” do SV no conjunto de treino. Diferente do n no somatório acima, o índice i no
    somatório abaixo já seleciona apenas os vetores de suporte. O somatório acima, percorre todo
    o conjunto de treino, e os exemplos que não são vetores de suporte possuem n = 0 e daí não
    são contabilizados.
    Note que a documentação é confusa, pois chama de “dual coefficients” para ai e depois indica
    que dual_coef_ armazena o produto li = yi ai.
    c) Usando a SVM escolhida e os dois primeiros exemplos z1 e z2 do conjunto de teste
    descrito em dataset_test.txt, compare o resultado numérico do “score” que você obtém para
    a função de decisão f(z) calculada pelos parâmetros do item b) anterior com os gerados pela
    função decision_function() da classe SVM do scikit-learn. Busque explicar eventuais
    discrepâncias.

In [3]:
import numpy as np
from sklearn import svm, datasets
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split

# Função para carregar o dataset personalizado
def load_custom_dataset():
    my_data = np.genfromtxt('dataset_train.txt', delimiter=',')
    X = my_data[:, :2]  # Assume que as duas primeiras colunas são features
    y = np.ravel(my_data[:, 2:])  # Assume que a terceira coluna é o target
    return X, y

# Carregar e preparar dados
X, y = load_custom_dataset()

# Dividir em treino e validação (80/20)
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)

# Normalização (baseada no treino)
scaler = StandardScaler()
X_train_norm = scaler.fit_transform(X_train)
X_val_norm = scaler.transform(X_val)

# Hiperparâmetros para tuning
C_values = [0.01, 1, 100]
gamma_values = [0.5, 1]

# Parte a) Tuning de hiperparâmetros
best_score = -1
best_svm = None
for C in C_values:
    for gamma in gamma_values:
        svm_rbf = svm.SVC(kernel='rbf', C=C, gamma=gamma, verbose=False)
        svm_rbf.fit(X_train_norm, y_train)
        score = svm_rbf.score(X_val_norm, y_val)
        print(f"C={C}, gamma={gamma}: Acurácia={score:.4f}")
        if score > best_score:
            best_score = score
            best_svm = svm_rbf

print("\nMelhor SVM:")
print(f"C={best_svm.C}, gamma={best_svm.gamma}, Acurácia={best_score:.4f}")

# Parte b) Descrição detalhada do melhor modelo
print("\nParâmetros do melhor modelo:")
print(f"b1) gamma = {best_svm.gamma}")
print(f"b2) Número de vetores de suporte por classe: {best_svm.n_support_}")
print(f"b3) Índices dos SVs no conjunto de treino: {best_svm.support_}")
print(f"b4) Coeficientes duais (λ * y):\n{best_svm.dual_coef_}")
print(f"b5) Bias (intercept_): {best_svm.intercept_[0]}")

# Parte c) Comparação manual vs decision_function
def manual_decision_function(z, svm):
    gamma = svm.gamma
    SVs = svm.support_vectors_
    dual_coef = svm.dual_coef_[0]
    b = svm.intercept_[0]
    total = 0
    for i in range(len(SVs)):
        x_i = SVs[i]
        kernel = np.exp(-gamma * np.linalg.norm(z - x_i)**2)
        total += dual_coef[i] * kernel
    return total + b

# Carregar dados de teste
test_data = np.genfromtxt('dataset_test.txt', delimiter=',')
X_test = test_data[:, :2]
y_test = np.ravel(test_data[:, 2:])
X_test_norm = scaler.transform(X_test)

# Comparar para os dois primeiros exemplos
z1, z2 = X_test_norm[0], X_test_norm[1]

f_z1_manual = manual_decision_function(z1, best_svm)
f_z1_sklearn = best_svm.decision_function([z1])[0]
print(f"\nComparação para z1:")
print(f"Manual: {f_z1_manual:.6f}, Scikit-learn: {f_z1_sklearn:.6f}")

f_z2_manual = manual_decision_function(z2, best_svm)
f_z2_sklearn = best_svm.decision_function([z2])[0]
print(f"\nComparação para z2:")
print(f"Manual: {f_z2_manual:.6f}, Scikit-learn: {f_z2_sklearn:.6f}")

C=0.01, gamma=0.5: Acurácia=0.0000
C=0.01, gamma=1: Acurácia=0.0000
C=1, gamma=0.5: Acurácia=1.0000
C=1, gamma=1: Acurácia=1.0000
C=100, gamma=0.5: Acurácia=1.0000
C=100, gamma=1: Acurácia=1.0000

Melhor SVM:
C=1, gamma=0.5, Acurácia=1.0000

Parâmetros do melhor modelo:
b1) gamma = 0.5
b2) Número de vetores de suporte por classe: [1 3]
b3) Índices dos SVs no conjunto de treino: [2 0 1 3]
b4) Coeficientes duais (λ * y):
[[-1.          0.44858894  0.44544123  0.10596983]]
b5) Bias (intercept_): 0.5513468251219731

Comparação para z1:
Manual: -0.303535, Scikit-learn: -0.303535

Comparação para z2:
Manual: 0.897922, Scikit-learn: 0.897922


Os resultados demonstram o processo de otimização de hiperparâmetros para uma SVM com kernel RBF, onde foram testadas combinações de valores para C (0.01, 1, 100) e gamma (0.5, 1). Observamos que:

**1. Desempenho por Configuração:**
- Para C=0.01 (com ambos gammas), a acurácia foi 0%, indicando que o modelo foi excessivamente regularizado e não conseguiu aprender padrões úteis nos dados.
- Todas as outras combinações (C=1 e C=100) alcançaram acurácia perfeita (100%), mostrando que esses valores permitiram um melhor ajuste aos dados.

**2. Melhor Modelo Selecionado:**
O modelo com C=1 e gamma=0.5 foi escolhido como o melhor, embora outras combinações tenham tido o mesmo desempenho. Essa escolha pode ser justificada por:
- Evitar overfitting (C=1 é menos propenso a sobreajuste que C=100)
- Manter boa generalização (gamma=0.5 oferece um equilíbrio na influência dos vetores de suporte)

**3. Características do Modelo Final:**
- Utiliza 4 vetores de suporte (1 da primeira classe e 3 da segunda)
- Os coeficientes duais mostram que o primeiro vetor de suporte tem maior peso (λ=-1.0)
- O bias positivo (0.551) indica um limiar de decisão deslocado

**4. Validação da Função de Decisão:**
A comparação entre os cálculos manuais e a função decision_function() do scikit-learn mostrou concordância perfeita para dois exemplos de teste:
- z1 foi classificado como negativo com score -0.303535
- z2 foi classificado como positivo com score 0.897922

**5. Implicações Práticas:**
- O modelo final tem excelente capacidade de generalização (acurácia 100%)
- A consistência nos cálculos valida a implementação
- A seleção de C=1 e gamma=0.5 sugere que:
  * Um C moderado foi suficiente para separar as classes
  * Um gamma intermediário capturou bem a complexidade dos dados sem sobreajuste
